app/models/user.rb

```diff
@@ -4,4 +4,14 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
+
+  validates :firstname, presence: true, length: { maximum: 50 }
+  validates :lastname, presence: true, length: { maximum: 50 }
+  validates :bio, length: { maximum: 300 }
+
+  has_many :posts
+  has_many :notifications
+  has_many :friends
```

Comment:
Friends relationship is very problematic. That's why I am going to save you some time:)

You might try to do a lot of tricks with associations, I recommend some reading of what is possible: http://guides.rubyonrails.org/association_basics.html#choosing-between-has-many-through-and-has-and-belongs-to-many.

However!

In this case user can make friend by:
- send request to someone and wait for acceptance
- OR receive request from someone and accept it.

That means that we would need to look for user id in `requestor_id` column but also in `friend_id`. And there is no way to specify foreign key as (column1 OR column2).

What will be the most clear and reasonable solution?
In **FriendRequest** model:

- add scope

```
scope :mutual_for_user, ->(user) {
    where("(requestor_id = :user_id OR friend_id = :user_id)", user_id: user.id)
  }
```

The you can use:

```
user = User.first
FriendRequest.mutual_for_user(user).accepted
```

In that case 🤔 if you would like to display a list of friends, you will have to take requestor and friend object and display the one who is not the current user.

------------------------------

There is another approach to friendships. We might create to rows for each friendship - so lets say we have user X and Y.

- row 1: requestor_id: X, friend_id Y
- row 2: friend_id: Y, requestor_id: X.

In that case: first row is created when request is sent and second when request is accepted.
We will need to remember to create second row every time the request is accepted.

But you will be able to define all `friends` relationship as 

```
has_many :friends, through: :friend_requests
```

-----------

Please give it a though and decide what do you prefer. I think that both approaches are equally good - we will just need to deal with different challenges.

Reply:
@talihomz  @SavaVuckovic some breaking news here! ;)

Actually you are going to need nice `friends` associations sooner or later.
It means that, we should use second approach - with 2 rows.

You can read about advantages here https://explainextended.com/2009/03/07/selecting-friends/. I really love this article.

You can also try this step by step guide https://dankim.io/mutual-friendship-rails/ which extends my suggestion.